compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -217,6 +240,188 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
+  extension (deps: ReverseDeps) def at(param: TypeParamRef): SimpleIdentitySet[TypeParamRef] =
+    val result = deps(param)
+    if null == result  // swapped operand order important since `==` is overloaded in `SimpleIdentitySet`
+    then SimpleIdentitySet.empty
+    else result
+
+  override def dependsOn(tv: TypeVar, except: TypeVars, co: Boolean)(using Context): Boolean =
+    def origin(tv: TypeVar) =
+      assert(!tv.isInstantiated)
+      tv.origin
+    val param = origin(tv)
+    val excluded = except.map(origin)
+    val qualifies: TypeParamRef => Boolean = !excluded.contains(_)
+    def test(deps: ReverseDeps, lens: ConstraintLens[List[TypeParamRef]]) =
+      deps.at(param).exists(qualifies)
+      || lens(this, tv.origin.binder, tv.origin.paramNum).exists(qualifies)
+    if co then test(coDeps, upperLens) else test(contraDeps, lowerLens)
+
+  /** Modify traversals in two respects:
+   *   - when encountering an application C[Ts], where C is a type variable or parameter
+   *     that has an instantiation in this constraint, assume the type parameters of
+   *     the instantiation instead of the type parameters of C when traversing the
+   *     arguments Ts. That can make a difference for the variance in which an argument
+   *     is traversed. Example constraint:
+   *
+   *         constrainded types: C[X], A
+   *         A >: C[B]
+   *         C := Option
+   *
+   *     Here, B is traversed with variance +1 instead of 0. Test case: pos/t3152.scala
+   *
+   *   - When typing a prefx, don't avoid negative variances. This matters only for the
+   *     corner case where a parameter is instantiated to Nothing (see comment in
+   *     TypeAccumulator#applyToPrefix). When determining instantiation directions in
+   *     interpolations (which is what dependency variances are for), it can be ignored.
+   */
+  private trait ConstraintAwareTraversal[T] extends TypeAccumulator[T]:
+
+    override def tyconTypeParams(tp: AppliedType)(using Context): List[ParamInfo] =
+      def tparams(tycon: Type): List[ParamInfo] = tycon match
+        case tycon: TypeVar if !tycon.isInstantiated => tparams(tycon.origin)
+        case tycon: TypeParamRef =>
+          entry(tycon) match
+            case _: TypeBounds => tp.tyconTypeParams
+            case tycon1 if tycon1.typeParams.nonEmpty => tycon1.typeParams
+            case _ => tp.tyconTypeParams
+        case _ => tp.tyconTypeParams
+      tparams(tp.tycon)
+
+    override def applyToPrefix(x: T, tp: NamedType): T =
+      this(x, tp.prefix)
+  end ConstraintAwareTraversal
+
+  private class Adjuster(srcParam: TypeParamRef)(using Context)
+  extends TypeTraverser, ConstraintAwareTraversal[Unit]:
+
+    var add: Boolean = compiletime.uninitialized
+    private val seen = util.HashSet[LazyRef]()
+
+    def update(deps: ReverseDeps, referenced: TypeParamRef): ReverseDeps =
+      val prev = deps.at(referenced)
+      val newSet = if add then prev + srcParam else prev - srcParam
```

Comment:
Is this an optimization or does removing entries instead of doing nothing when `add == false` actually matter for correctness?

Reply:
Removing entries from `prev` could be necessary fro correctness. Otherwise we have stale backwards references to parameters that do not correspond to any forward reference. Whether this is a problem globally I am  not sure, but better not risk it. 

Removing the entry array from the deps set (in line 324) is a pure optimization. It also avoids printing junk.